scalafix-reflect/src/main/scala-3/scalafix/internal/reflect/RuleCompiler.scala

```diff
@@ -0,0 +1,93 @@
+package scalafix.internal.reflect
+
+import scalafix.internal.reflect.exception.CompilationException
+
+import java.io.File
+import java.nio.file.Paths
+import scala.meta.io.AbsolutePath
+
+import dotty.tools.dotc.Compiler
+import dotty.tools.dotc.Run
+import dotty.tools.dotc.util.SourceFile
+import dotty.tools.dotc.core.Contexts.NoContext.given_Context
+import dotty.tools.dotc.reporting.StoreReporter
+
+import scala.reflect.io.VirtualDirectory
+import scala.reflect.io.AbstractFile
+import scala.reflect.internal.util.AbstractFileClassLoader
+
+import metaconfig.Configured
+import metaconfig.Input
+
+class RuleCompiler(
+    classpath: String,
+    target: AbstractFile = new VirtualDirectory("(memory)", None)
+) {
+  // we are not configuring any settings!
+  // private val settings = new Settings()
+  // settings.deprecation.value = true // enable detailed deprecation warnings
+  // settings.unchecked.value = true // enable detailed unchecked warnings
+  // settings.outputDirs.setSingleOutput(target)
+  // settings.classpath.value = classpath
+
+  private lazy val reporter: StoreReporter = new StoreReporter()
+  private val compiler: Compiler = new Compiler()
+  private val classLoader: AbstractFileClassLoader =
+    new AbstractFileClassLoader(target, this.getClass.getClassLoader)
+
+  @throws(classOf[CompilationException])
+  def compile(input: Input): Configured[ClassLoader] = {
+    val run: Run = compiler.newRun
+
+    run.compileSources(
+      List(
+        new SourceFile(
+          dotty.tools.io.AbstractFile.getFile(input.path),
+          input.chars
+        )
+      )
+    )
+
+    if (reporter.hasErrors) {
+      val cause = reporter.allErrors { 0 }.getCause()
+      if (reporter.errorCount > 1) {
+        val msg = reporter.allErrors { 0 }.getMessage()
+
+        throw new CompilationException(msg, cause)
+      }
+      throw new CompilationException(
+        "Multiple compilation errors occurred",
+        cause
+      )
+    }
+
+    Configured.Ok(classLoader)
+  }
+}
+object RuleCompiler {
+
+  def defaultClasspath: String = {
+    defaultClasspathPaths.mkString(File.pathSeparator)
+  }
+
+  def defaultClasspathPaths: List[AbsolutePath] = {
+    val classLoader = ClasspathOps.thisClassLoader
+    val paths = classLoader.getURLs.iterator.map { u =>
+      if (u.getProtocol.startsWith("bootstrap")) {
+        import java.nio.file._
+        val stream = u.openStream
+        val tmp = Files.createTempFile("bootstrap-" + u.getPath, ".jar")
+        try {
+          Files.copy(stream, tmp, StandardCopyOption.REPLACE_EXISTING)
+        } finally {
+          stream.close()
+        }
+        AbsolutePath(tmp)
+      } else {
+        AbsolutePath(Paths.get(u.toURI))
+      }
+    }
+    paths.toList
+  }
```

Comment:
this didn't change from the original scala2 source, it should be refactored out and reused across scala versions probably

Reply:
done, in `RuleCompilerClasspath`